### ISSUE TITLE
IOU thesold in inference VS  IOU thesold in evaluation

### ISSUE THREAD

**Issue by IgorGoldberg** [ultralytics/ultralytics#8634]
1) I guess the IOU thesold in inference and the IOU thesold in evaluation are not the same.
In inference, the IOU  is calculated against other detections while in evaluation it is calculated against GT
Am I right?

2)If yes, here is a question:
While evaluating there is only one IOU threshold parameter:
metrics  = model.val(data = path_to_data, imgsz=my_im_size, **iou = 0.5** ) 
I can guess it refers to the thresholding against the GT, but there is also an inference that is done as a part of the evaluation process, and I wonder which IOU threshold is this inference done with. Can I change it too?

**Comment by glenn-jocher** [comment]
@IgorGoldberg hello! 😊 You're on the right track with your understanding.

1) Yes, you're correct. During inference, the IOU threshold is used to filter out overlapping detections (Non-Maximum Suppression, NMS), ensuring each detected object is unique. In evaluation, the IOU threshold helps determine if a detection matches a ground truth (GT) object based on their overlap.

2) For the evaluation process, the IOU threshold you mentioned (`iou=0.5`) indeed refers to the threshold against the GT during the validation phase. The inference part of the evaluation uses the NMS IOU threshold defined in the model or the training configuration. If you're looking to adjust the NMS IOU threshold for inference during evaluation, you would typically need to modify this in your model's configuration file or directly in the code if the API allows.

For example, when initializing your model for training or inference, you might have the option to set the NMS IOU threshold like so (hypothetically, as the exact implementation may vary):

```python
model = YOLO('yolov8n.pt', nms_iou=0.4)
```

However, this might not directly apply to the `model.val()` method as you described. The ability to change the inference IOU threshold during evaluation depends on the specific implementation of the YOLOv8 API you're using. It's always a good idea to check the official documentation or source code for the most accurate information.

Hope this helps clarify things a bit! Let me know if you have more questions.

**Comment by IgorGoldberg** [comment]
Hello Glenn! Thank you.
Now I am a bit confused though I am on the right path :)

1)
Here:
https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation  **(evaluation)**
https://docs.ultralytics.com/modes/predict/#inference-arguments  **(inference)**
In both documents above no "nms_iou"  argument is mentioned, they both have same one called "iou".
So please clarify me what dod you mean by "nms_iou".
2)
Here 
https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters  **(trainig)**
no such argument at all - not the "iou" not the "nms_iou", Actually I ddint expect them to be here, but you are saying 
"defined in the model or the training configuration"... PLease clarify this point to me also.

------
In case you need , here is more info:
I train via Python scripts in pycharm
My training code looks like this:
train.py

`
    from ultralytics import YOLO
    model = YOLO('yolov8n.pt')  
    data_yaml_file_path = 'some/path/data.yaml'
    where_to_store_the_model = r'another/path/' 
    results = model.train(data= data_yaml_file_path,
                          epochs=500,
                          imgsz=848,
                          rect = True, #works only if all images are of a same size
                          batch = 6,
                          patience = 50,
                          val=True,
                          name = where_to_store_the_model,
                          )
`

**and the inference code looks like this:**
inf.py

def main():
`
    from ultralytics import YOLO
    model = YOLO('some/other/path')
    source = 'some/path'
    
    results = model.predict(source,
                classes = [0, 1, 2, 3]
                show_boxes = True,
                save = True,
                show_labels = False,
                show_conf = False,
                conf = 0.1,
                save_txt = True,
                save_crop = False,
                line_width = 2,
                **iou = 0.5,** 
                save_conf = False 
                )
`







**and the val code looks like this:**
evaluate.py:
`
from ultralytics import YOLO
model = YOLO('some/path/last.pt')  
path_to_data = 'some/other/path/data.yaml' 
# Validate the model
metrics = model.val(data = path_to_data, imgsz=848, **iou = 0.5** )
# metrics.box.map    # map50-95
metrics.box.map50  # map50
# metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category
print('done')`

**Comment by glenn-jocher** [comment]
@IgorGoldberg hey there! 😊 Thanks for reaching out and for the detailed info!

Regarding your confusion:

1) The term "nms_iou" I mentioned was to differentiate between the IOU used during Non-Maximum Suppression (NMS) in inference and the IOU used for matching detections to ground truth in evaluation. It seems there was a bit of a mix-up on my part, as the documentation indeed refers to both simply as "iou". My apologies for the confusion! In practice, the "iou" parameter you see in both the validation and prediction settings serves the purpose I described - for validation, it's about matching to ground truth, and for inference, it's part of the NMS process.

2) For training, while "iou" or "nms_iou" isn't directly specified as an argument in your script, these values are often set within the model's configuration or the YAML file defining the dataset and hyperparameters. They're part of the model's internal settings that influence how it learns to make detections and aren't typically adjusted per training run but can be tweaked for fine-tuning or experimentation.

Your training, inference, and evaluation scripts look solid! If you're looking to adjust the NMS IOU threshold specifically for inference, this would typically be done by modifying the model's configuration file or directly in the code if the API allows. However, based on the current documentation and your scripts, it seems like using the "iou" parameter in your inference and evaluation scripts is the correct approach.

I hope this clears things up a bit! If you have any more questions, feel free to ask. Happy coding! 🚀

**Comment by wliu20** [comment]
Hello, I have the same question about the param 'iou'.
In the default.yaml file it is commented as nms_iou, and I guess whether in evaluation or inference it's always been NMS IOU,  that is used during Non-Maximum Suppression (NMS). But one more question for me is how to change the iou threshold against the GT  for evaluation.

**Comment by glenn-jocher** [comment]
@wliu20 hello! 😊 Great question! The 'iou' parameter you see in the `default.yaml` indeed refers to the NMS IOU used during inference for Non-Maximum Suppression. This helps in filtering out overlapping detections to ensure each detected object is unique.

For changing the IOU threshold against the Ground Truth (GT) during evaluation, this is typically specified when you run the evaluation. For example, in Python:

```python
metrics = model.val(data='path/to/data.yaml', imgsz=640, iou=0.5)
```

Here, `iou=0.5` sets the IOU threshold for matching detections to GT during evaluation. Unfortunately, if you're looking to adjust this directly in the `default.yaml` or another configuration file, such an option isn't directly available. It's usually passed as an argument in the evaluation method as shown above.

Hope this helps! Let me know if you have any more questions. Happy to assist further!

**Comment by mariamiguel01** [comment]
Dear @glenn-jocher,
I have been using YOLO for training and was performing the **validation** (model.val) in order to extract metrics . 

My goal was to test out **25 combinations of IOU and confidence thresholds** so that I could access how they make the evaluation metrics change (F1, Recalll and Precision). 

To my surprise, every **confusion matrix runned with the same confidence** had exactly the same look with the **exact same TP FP and FN**, indicating that **IOU is not used as the matching thresholding of a bounding box with its GT**.  Instead, it is used to **control IOU NMS** as described in your documentation( https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation). In my case, I was using a max_det=1,  so changing the IOU (IOU NMS) did not affect my results, because I already had only one bounding box being predicted per frame. 

Since I came to that conclusion I have been trying to understand **where the threshold for the IOU with the GT is defined** and I can't seem to find it. Then I came across this forum states exactly the opposite of what is described in the documentation and also of my own conclusions. Is that true? Is it possible to actually control the IOU with the GT? Or is it only possible to control the IOU_NMS?

Thank you so much!

**Comment by glenn-jocher** [comment]
@mariamiguel01 thank you for your detailed analysis! You're correct that the `iou` parameter in validation controls the NMS threshold, not the GT matching. The IOU threshold for matching detections to ground truth is fixed at **0.45** in the underlying metrics calculation ([source](https://docs.ultralytics.com/reference/utils/metrics/#confusionmatrix)). This value isn't currently exposed as a user-configurable parameter in `model.val()`, but we appreciate the feedback and will consider making this adjustable in future updates. For now, you'd need to modify the `iou_thres` parameter directly in the `ConfusionMatrix` class for custom GT matching thresholds.
